### PR TITLE
rhit: fix build and test for Linux

### PR DESCRIPTION
**Pull request by dawidd6**
- [ ] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [ ] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

**Comment by dawidd6** [comment]
@Homebrew/linux what do we do here with tests that require /dev/pts/0?

**Comment by danielnachun** [comment]
Up until now, we've just been disabling them on Linux.  We've been doing the same for tests which require X server or Wayland as well.  We have previous discussed that we should probably set up proper PTY and X/Wayland interfaces in our Docker CI so that these tests can actually be used, but I personally don't know much about how to do that.

**Comment by cho-m** [comment]
The error here may be Ruby behavior
`Errno::EIO: Input/output error @ io_fread - /dev/pts/0`

https://ruby-doc.org/stdlib-2.6.3/libdoc/pty/rdoc/PTY.html
```rb
# The result of read operation when pty slave is closed is platform
# dependent.
ret = begin
        master.gets     # FreeBSD returns nil.
      rescue Errno::EIO # GNU/Linux raises EIO.
        nil
      end
```

Not sure what is best way to deal with though.

I tried on local Docker with same test and hit failure.
I hacked around the exception (though I am not sure if correct) with something like:
```rb
      output = ""
      PTY.open do |master, slave|
        master.winsize = [80, 130]
        slave.raw!
        pid = spawn(cmd, :out=>slave)
        slave.close

        begin
          master.each_line { |line| output += line }
        rescue Errno::EIO
        ensure
          Process.wait pid
        end
      end
```

**Comment by cho-m** [comment]
Also confirmed following worked for me.
Once again, need someone with more Ruby experience to clean up.
```rb
      output = ""
      PTY.spawn("#{bin}/rhit --silent-load --length 0 --color no access.log") do |r, _w, _pid|
        r.winsize = [80, 130]
        begin
          r.each_line { |line| output += line.gsub(/\r?\n/, "\n") }
        rescue Errno::EIO
        end
      end
```

File: Formula/rhit.rb [hunk 1]
```diff
@@ -31,27 +29,32 @@ def install
...
-          │ 2xx │ 3xx │4xx │5xx │
-          ├─────┼─────┼────┼────┤
-          │79.1%│14.9%│1.2%│4.8%│
-          └─────┴─────┴────┴────┘
-        EOS
+        begin
+          r.each_line { |line| output += line.gsub(/\r?\n/, "\n") }
+        rescue Errno::EIO
+          # GNU/Linux raises EIO when read is done on closed pty slave
+        end
```
Comment (cho-m) [inline Formula/rhit.rb:39]: Added changes to see how CI behaves.
On local macOS and docker Linux, the test passes.

**Comment by cho-m** [comment]
> We have previous discussed that we should probably set up proper PTY and X/Wayland interfaces

@danielnachun Some of the PTY issues are probably similar to this formula where the Ruby code needs to be updated to support platform-dependent behavior.

X/Wayland will probably be a bit trickier.

For X, maybe adding something like `xvfb` as dependency and forking process during test. Haven't tried it myself and I don't know how it will behave within Docker. Also, not available in Homebrew right now.

**Comment by danielnachun** [comment]
> > We have previous discussed that we should probably set up proper PTY and X/Wayland interfaces
> 
> @danielnachun Some of the PTY issues are probably similar to this formula where the Ruby code needs to be updated to support platform-dependent behavior.

Fantastic work looking into this more deeply.  It does seem that `Errno:EIO` is the expected output when a PTY connection is closed in Linux, and we just need to rescue the error as you've done here.  

> X/Wayland will probably be a bit trickier.
> 
> For X, maybe adding something like `xvfb` as dependency and forking process during test. Haven't tried it myself and I don't know how it will behave within Docker. Also, not available in Homebrew right now.

That sounds like a good starting point.  I guess we would start by adding `xvfb` as formula and seeing if it behaves in a reasonable way in CI.

File: Formula/rhit.rb [hunk 2]
```diff
@@ -31,27 +29,32 @@ def install
...
-          ┌─────┬─────┬────┬────┐
-          │ 2xx │ 3xx │4xx │5xx │
-          ├─────┼─────┼────┼────┤
-          │79.1%│14.9%│1.2%│4.8%│
-          └─────┴─────┴────┴────┘
-        EOS
+        begin
+          r.each_line { |line| output += line.gsub(/\r?\n/, "\n") }
+        rescue Errno::EIO
+          # GNU/Linux raises EIO when read is done on closed pty
```
Comment (cho-m) [inline Formula/rhit.rb:38]: Before merge, I wanted to check with someone more familiar with Ruby to see what is best way to do this. There are a couple other formulae that need to similar fix, which means the code chosen will probably be copied elsewhere.

**Review by cho-m — APPROVED** [review]

**Comment by cho-m** [comment]
Don't think it is worth waiting any longer. Will just merge as build/test is all passing.

**Pull request merged**